### ISSUE TITLE
events subtask: converting marketing spreadsheet to new `event.js`

### ISSUE THREAD

**Issue by mattxwang** [uclaacm/website#281]
This issue is a subtask of #115 - building our events page.

## Overview

For our MVP, we won't be building any sort of API endpoint; thus, to load in the event data, we'll need to generate a unified list of events from our existing data sources. There are two good data sources for this:

1. the marketing spreadsheet
2. the google calendar

I want to first explore using the former (rather than the latter), since a successful implementation completely removes our dependency on the current Google Calendar (note: need to check with our Zapier integration?).

## Deliverable Requirements/non-requirements

We need to have a script that converts items from the marketing spreadsheet to the events JSON format. 

Things that the deliverable does **not need**:

* the events do not need to be in any specific order
* all-day events can be tabled until later or evaluated on a case-by-case basis
* limited edge cases can be manually corrected
* we do not need to deal with multi-committee or cross-org events (as a special case)
* while there's a preference for JS, language dependency is not strict - this is a monkeypatch solution anyways.

Also, note that **we can make adjustments to the marketing spreadsheet to make our lives easier**, since it has not been created. Turnaround is tight on this :( 

## Strategy

Assuming we're using the marketing spreadsheet, we then need a workflow that converts from it to the current data structure. Looking at the CSV format for last year's spreadsheet, we will need to create two separate scripts: one for recurring events, and one for one-offs.

For one-off events, we should look to map the CSV headers to the current JSON structure (see below). We should be able to directly map the `title`, `committee` (or a simple lowercase + slug), and `description`. Some other fields are more complicated:

* `id` should be auto-generated
* `allDay` can be `false` by default; we run few all-day events
* `start` and `end` needs to be done by machine-parsing the date and time, which probably involves delimiting the string and then running it through JS `Date()`. A couple of extra notes:
    * we should assume the CSV is in the PST timezone, and encode the timezone in the date object
    * we currently have [`moment.js`](https://momentjs.com/) installed in the branch! check it out :)
    * recall that JS Unix Timestamp is in milliseconds and not seconds!
* `image` and `alt` are very hard to do properly, since people tend to link drive images incorrectly. Your call on how to do this.
* `links` relies on us figuring out if the event is online or in-person. we then need to check if the location thing is a zoom URL

For recurring events, the idea is similar. However, instead of mapping the CSV entry to one event, we want to turn the time range into some sort of Date array, and then create one Event per set of Dates.

## Current JSON structure

The schema for an event is as follows (in pseudo-TS)

```ts
type Event {
  id: number, // this is used for housekeeping by our JS library
  title: string, // title of the event
  allDay: boolean, // whether or not the event spans the entire day
  start: Date, // start time for the event; if the event is all day, only the day is used
  end: Date, // (non-inclusive) end time for the event; if the event is all day, only the day is used
  committee: Enum<string>?, // a string (of `ai`, `cyber`, `design`, `hack`, `icpc`, `studio`, `teach-la`, `w`) to note which committee runs the event. if none is provided, we default to board branding.
  location: string, // a string (or, JSX) that goes into the "location" line
  description: string, // description for the event
  image: string?, // a path to a URL for an image. if non is provided, we use a default
  alt: string?, // alt text for an image. if non is provided, we use a default, though this is highly discouraged
  links: Array<ResourceLink>?, // an optional array of resource links (defined below)
}
```

We define a `ResourceLink` as:

```ts
type ResourceLink {
  text: string, // text for link
  href: string, // hyperlink for link; can be absolute or relative
  ext: boolean, // whether or not the link is external (and thus should be opened in a new tab)
}
```

Here's an example of an all-day event not run by a committee, with no image or alt text and an external link:

```js
const events = [
  {
    id: 1,
    title: 'CS Welcome Day',
    allDay: true,
    start: new Date(2021, 8, 21),
    end: new Date(2021, 8, 22),
    committee: 
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc quam orci, aliquet eget mollis at, facilisis porta odio. Cras mauris nisl, suscipit nec posuere semper, euismod sit amet nibh. Nulla facilisi. Praesent quam lacus, vulputate vel condimentum eget, congue ac justo.',
    links: [
        { text: 'Event Website', href: 'https://www.cs.ucla.edu/cs-welcome-day/', ext: true },
    ],
  },
];
```

Here's an example of a timed event run by a committee, with an image, alt text, and no external link:

```js
const events = [
  {
    id: 7,
    title: 'Studio Fall GM',
    allDay: false,
    start: new Date(2021, 9, 6, 19),
    end: new Date(2021, 9, 6, 21),
    committee: 'studio',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc quam orci, aliquet eget mollis at, facilisis porta odio. Cras mauris nisl, suscipit nec posuere semper, euismod sit amet nibh. Nulla facilisi. Praesent quam lacus, vulputate vel condimentum eget, congue ac justo.',
    image: '/images/events/f21-fall-gm.png',
    alt: 'a banner that says "ACM at UCLA: Fall GM, Monday, September 27th from 6-8 PM in the Court of Sciences". Features graphical elements representing each of ACM\'s committees: ACM Hack, ACM W, ACM AI, ACM ICPC, ACM Cyber, ACM Studio, ACM Teach LA, and ACM Design.',
  },
];
```

## One-off Event Workflow - Marketing Spreadsheet

The CSV has the resulting properties:

1. Committee -> subset of valid strings
2. Event/Workshop Title -> any string
3. Date -> datestring (machine-parseable)
4. Time -> time range (machine-parseable, if delimited by `-`
5. Location/Zoom Link -> string; may have to programmatically check if it's a zoom event?
6. Description - string
7. Facebook Link - url
8. Newsletter Banner - URL or "Use Facebook event banner", which is hard to machine parse

## Recurring Event Workflow - Marketing Spreadsheet

The CSV has the resulting properties:

1. Committee -> subset of valid strings
2. Event/Workshop Title -> any string
3. RANGE of weeks -> currently a string (but we should ask cprez to implement this in some sort of computer-readable date?)
4. Time -> a time string, notably without the day of the week encoded; we can ask cprez to change this
5. Location/Zoom Link -> string; may have to programmatically check if it's a zoom event?
6. Description - string
7. Facebook Link - url
8. Newsletter Banner - URL or "Use Facebook event banner", which is hard to machine parse

**Comment by mattxwang** [comment]
Updated w/ the new `location` field I've introduced!

**Comment by mattxwang** [comment]
Late, but closed by #272.
